### ISSUE TITLE
Barcode trimming with a list of barcodes for specific location

### ISSUE THREAD

**Issue by ghoshal** [marcelm/cutadapt#336]
Hi,

I am trying to look for a list of barcodes (96 in total) at the exact same location (bases 89-96) in the read 2 of a paired end fastq. Is there a way to do this using cutadapt? I want matched paired end fastq files that match the barcodes in the barcode file in those bases and remove the rest that don't.

Thanks,

Bibaswan

**Comment by marcelm** [comment]
This isn’t really supported well at the moment. I think one option is to provide your 96 barcode sequences as anchored 5' adapters like this: `-G ^N{88}ACGTACGT`. The `N{88}' means “any sequence, as long as it is 88 nt long”. Better yet, put the 96 sequences into a FASTA file where you write the sequences (including the `^N{88}`), and then tell Cutadapt to use that file with `-G file:barcodes.fasta`.

This will be quite slow as Cutadapt tries to match every 96 of the “adapters” in turn. I plan to improve this use case: It could instead look at positions 89-96 in the read and then decide immediately which of barcodes it sees. However, this is not implemented at the moment.

**Comment by marcelm** [comment]
A note to anyone trying this: The error rate should also be adjusted as long as issue #360 is open.

**Comment by soluna1** [comment]
Hi,
Wanting to do something similar, i.e. to allow '3 adapters to occur at the end but also until a certain number of bases to the left (but not all of them) in order to avoid trimming the whole read. 

I thought that this could work, 
-g 'GTACACACCGCCCGTCA;e=0.12' -a 'CCGTAGGTGAACCTGCAGAAGGATCA**N{0,22}X**;e=0.07'
-G 'TGATCCTTCTGCAGGTTCACCTACGG;e=0.12' -A 'TGACGGGCGGTGTGTAC**N{0,22}X**;e=0.07'

but I've got this error
utadapt: error: invalid literal for int() with base 10: '0,22'

So, I think Cutadapt doesn't accept regex coding. is there any other way to say "find the 3' adapter at the end, but also a maximum of 22 bp to the left?"

thanks in advance 

Soluna

**Comment by marcelm** [comment]
Hi,

just a note before answering your actual question: Cutadapt by default trims *at most one* adapter from each read, so if you use `-g` and `-a` together as above, Cutadapt trims the 5' adapter *or* the 3' adapter but never both. If indeed you want to trim both, you need `--times=2` or, preferably, use a linked adapter (see docs).

Regarding your question: If I understand it correctly, you should just write `N{22}X` instead. With a non-internal 3' adapter, partial overlaps are still allowed, so the sequence is allowed to occur anywhere in the read as long as it is followed by at most 22 other bases.

**Comment by soluna1** [comment]
Hi,

thanks for your answer, Yes, I can confirm that --times = 2 is included in the parameters.
Ok. I will do that, I thought in normal regex syntax, but if it allows 0 to 22 N bases, that's great.

**Comment by marcelm** [comment]
Yes, I based the notation off regular expressions, but full regular expressions are not actually supported. Note that `N{22}` will only only 0 to 22 N bases because it happens to be at the end of the adapter specification string. It is just a shorthand for writing `N` 22 times.
